# Employee Salary Prediction - Model Training

## Objective

The objective of this notebook is to build, train, and evaluate multiple machine learning regression models for predicting employee salaries.

### Models Used

- Linear Regression
- Decision Tree Regressor
- Random Forest Regressor

### Evaluation Metrics

- Mean Absolute Error (MAE)
- Mean Squared Error (MSE)
- Root Mean Squared Error (RMSE)
- R² Score

### 1. Import Required Libraries

In [1]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LinearRegression
from sklearn.metrics import (mean_absolute_error, mean_squared_error, r2_score)
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
import pandas as pd
import numpy as np

### 2. Load the Cleaned Dataset

In [3]:
data = pd.read_csv("../data/employees_dataset_cleaned.csv")
df = pd.DataFrame(data)

In [4]:
df

,job_title,experience_years,education_level,skills_count,industry,company_size,location,remote_work,certifications,salary
0,AI Engineer,10,Bachelor,2,Healthcare,Medium,India,Hybrid,2,109413
1,Data Analyst,5,Bachelor,17,Telecom,Small,Australia,No,0,93764
2,Frontend Developer,18,PhD,4,Media,Medium,Singapore,No,1,148123
3,Business Analyst,19,PhD,13,Retail,Medium,Canada,Yes,0,189123
4,Product Manager,15,Bachelor,7,Manufacturing,Large,Sweden,Yes,0,165069
...,...,...,...,...,...,...,...,...,...,...
249995,Software Engineer,17,PhD,2,Telecom,Enterprise,India,No,1,127791
249996,Frontend Developer,20,PhD,7,Telecom,Startup,Remote,No,2,154593
249997,Business Analyst,1,Bachelor,12,Retail,Enterprise,India,Yes,0,75988
249998,Data Scientist,0,High School,2,Consulting,Small,Sweden,Hybrid,5,90467


## 3. Separate Features and Target Variable

The target variable for this project is **Salary**, while all remaining columns are used as input features.

In [5]:
x = df.drop("salary", axis=1)

In [6]:
y = df["salary"]

### 4. Identify Numerical and Categorical Features

To apply appropriate preprocessing techniques, the dataset is divided into numerical and categorical features.

In [7]:
num_features = ["experience_years", "skills_count", "certifications"]

In [8]:
cat_features = ["job_title", "education_level", "industry", "company_size", "location", "remote_work"]

In [9]:
print(x.head())

            job_title  experience_years education_level  skills_count  \
0         AI Engineer                10        Bachelor             2   
1        Data Analyst                 5        Bachelor            17   
2  Frontend Developer                18             PhD             4   
3    Business Analyst                19             PhD            13   
4     Product Manager                15        Bachelor             7   

        industry company_size   location remote_work  certifications  
0     Healthcare       Medium      India      Hybrid               2  
1        Telecom        Small  Australia          No               0  
2          Media       Medium  Singapore          No               1  
3         Retail       Medium     Canada         Yes               0  
4  Manufacturing        Large     Sweden         Yes               0  


In [10]:
print(y.head())

0    109413
1     93764
2    148123
3    189123
4    165069
Name: salary, dtype: int64


In [11]:
print(x.dtypes)

job_title           object
experience_years     int64
education_level     object
skills_count         int64
industry            object
company_size        object
location            object
remote_work         object
certifications       int64
dtype: object


In [12]:
SimpleImputer(strategy="median")

SimpleImputer(strategy='median')

### 5. Split Dataset into Training and Testing Sets

The dataset is divided into training and testing subsets using an 80:20 ratio.

- Training Set: 80%
- Testing Set: 20%

In [13]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [14]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(200000, 9)
(50000, 9)
(200000,)
(50000,)


## 6. Data Preprocessing Pipeline

### Numerical Pipeline
Numerical features are:

- Imputed using the median strategy
- Standardized using StandardScaler

In [15]:
num_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

### Categorical Pipeline
Categorical features are encoded using OneHotEncoder to convert text values into numerical representations.

In [16]:
# Categorical Pipeline
cat_pipeline = Pipeline([
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

### Combine Pipelines using ColumnTransformer

In [17]:
# Creating Full Pipeline
preprocessor = ColumnTransformer([
    ("num", num_pipeline, num_features),
    ("cat", cat_pipeline, cat_features)
])

### 7. Transform Training and Testing Data

In [18]:
x_train_prepared = preprocessor.fit_transform(x_train)

In [19]:
print(x_train_prepared.shape)

(200000, 48)


In [20]:
x_test_prepared = preprocessor.transform(x_test)

In [21]:
print(x_test_prepared.shape)

(50000, 48)


## Train Model

### 8. Linear Regression
Linear Regression is used as the baseline model for salary prediction.

In [57]:
lin_reg = LinearRegression()

In [58]:
lin_reg.fit(x_train_prepared, y_train)

LinearRegression()

In [59]:
salary_predictions = lin_reg.predict(x_test_prepared)

In [60]:
print(salary_predictions)

[172850.84694068  89234.51854182  63791.55134444 ... 165227.44895727
 151849.72558748 111755.77171383]


### Model Evaluation
The model is evaluated using:

- MAE
- MSE
- RMSE
- R² Score

In [61]:
mae = mean_absolute_error(y_test, salary_predictions)
mse = mean_squared_error(y_test, salary_predictions)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, salary_predictions)

In [62]:
print("Mean Absolute Error :", mae)
print("Root Mean Square Error :", rmse)
print("R² :", r2)

Mean Absolute Error : 5436.096958649476
Root Mean Square Error : 7125.522920575375
R² : 0.9634690226760201


### 9. Decision Tree Regressor
Decision Trees learn hierarchical decision rules from the training data.

In [63]:
dec_reg = DecisionTreeRegressor()
dec_reg.fit(x_train_prepared, y_train)

DecisionTreeRegressor()

In [64]:
salary_predictions = dec_reg.predict(x_test_prepared)

In [65]:
print(salary_predictions)

[165943.  85794.  51837. ... 153376. 140999. 118571.]


#### Evaluate Dicision Tree Regressor Model

In [66]:
mae = mean_absolute_error(y_test, salary_predictions)
mse = mean_squared_error(y_test, salary_predictions)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, salary_predictions)

In [67]:
print("Mean Absolute Error :", mae)
print("Root Mean Square Error :", rmse)
print("R² :", r2)

Mean Absolute Error : 7280.98978
Root Mean Square Error : 9233.110925411325
R² : 0.9386628045481528


### 10. Random Forest Regressor
Random Forest combines multiple decision trees to improve prediction accuracy and reduce overfitting.

In [33]:
random_forest_reg = RandomForestRegressor(n_estimators=50, n_jobs=-1, random_state=42)

In [34]:
random_forest_reg.fit(x_train_prepared, y_train)

RandomForestRegressor(n_estimators=50, n_jobs=-1, random_state=42)

In [35]:
salary_predictions = random_forest_reg.predict(x_test_prepared)

In [36]:
print(salary_predictions)

[167861.68  92881.16  70176.   ... 155343.02 148852.22 120405.26]


In [52]:
comparison.to_csv("../data/salary_predictions.csv",index=False)

#### Evaluate Random Forest Regressor

In [37]:
mae = mean_absolute_error(y_test, salary_predictions)
mse = mean_squared_error(y_test, salary_predictions)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, salary_predictions)

In [38]:
print("Mean Absolute Error :", mae)
print("Root Mean Square Error :", rmse)
print("R² :", r2)

Mean Absolute Error : 5048.546023452857
Root Mean Square Error : 6364.544987946304
R² : 0.9708551026755426


## 11. Model Comparison

Performance of all trained models is compared using common regression metrics.

In [40]:
results = pd.DataFrame({
    "Model": ["Linear Regression","Decision Tree","Random Forest"],
    "MAE": [5436.09,7291.96,5048.55],
    "RMSE": [7125.52,9251.13,6364.54],
    "R2 Score": [0.96346,0.93842,0.97085]
})
results

,Model,MAE,RMSE,R2 Score
0,Linear Regression,5436.09,7125.52,0.96346
1,Decision Tree,7291.96,9251.13,0.93842
2,Random Forest,5048.55,6364.54,0.97085


## 12. Best Performing Model

Based on the evaluation metrics, the **Random Forest Regressor** achieved the best overall performance among all the trained models.

### Performance Summary
| Metric | Random Forest |
|---------|--------------:|
| Mean Absolute Error (MAE) | **5048.55** |
| Root Mean Squared Error (RMSE) | **6364.54** |
| R² Score | **0.9709** |

### Conclusion

The Random Forest Regressor outperformed both Linear Regression and Decision Tree Regressor.

- ✅ Achieved the **highest R² Score (0.9709)**, indicating that it explains approximately **97% of the variance** in employee salaries.
- ✅ Produced the **lowest Mean Absolute Error (MAE)**, meaning its predictions are, on average, closest to the actual salaries.
- ✅ Recorded the **lowest Root Mean Squared Error (RMSE)**, demonstrating better overall prediction accuracy and fewer large prediction errors.

Therefore, the **Random Forest Regressor** is selected as the **best baseline model** for this project and will be used in the next phase for **hyperparameter tuning using GridSearchCV** to further improve its predictive performance.

In [47]:
comparison = pd.DataFrame({"Actual Salary": y_test.values, "Predicted Salary": salary_predictions})

In [51]:
comparison["Difference"] = (comparison["Actual Salary"] -comparison["Predicted Salary"])
comparison["Percentage Error"] = (abs(comparison["Difference"])/comparison["Actual Salary"]) * 100
comparison.head(15)

,Actual Salary,Predicted Salary,Difference,Percentage Error
0,164009,167861.68,-3852.68,2.349066
1,79594,92881.16,-13287.16,16.693670
2,74090,70176.00,3914.00,5.282764
3,177193,162563.08,14629.92,8.256489
4,120012,115070.18,4941.82,4.117772
5,163369,163882.26,-513.26,0.314172
6,111889,110474.78,1414.22,1.263949
7,75418,73300.76,2117.24,2.807340
8,103067,96603.02,6463.98,6.271629
9,190692,196073.36,-5381.36,2.822017


#### Hyper Parameter Tuning of Random Forest Regressor (Model Optimization)

In [22]:
rf = RandomForestRegressor(random_state=42)

In [23]:
param_grid = {
    "n_estimators": [50, 100, 150],
    "max_depth": [10, 20, None],
    "min_samples_split": [2, 5],
    "min_samples_leaf": [1, 2]
}

In [24]:
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5,scoring="r2", n_jobs=-1)

In [25]:
estimator = rf

In [26]:
param_grid = param_grid

In [ ]:
grid_search.fit(X_train_prepared, y_train)

In [28]:
print(grid_search.best_params_)

In [ ]:
print(grid_search.best_score_)

In [ ]:
predictions = best_model.predict(X_test_prepared)

In [ ]:
mae = mean_absolute_error(y_test, predictions)
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, predictions)

print("MAE :", mae)
print("RMSE:", rmse)
print("R² :", r2)